framework/src/headers/ip/v6/icmp/ndp_options.rs

```diff
@@ -0,0 +1,229 @@
+   Description
+                     The Source Link-Layer Address option contains the
+                     link-layer address of the sender of the packet.  It
+                     is used in the Neighbor Solicitation, Router
+                     Solicitation, and Router Advertisement packets.
+
+                     The Target Link-Layer Address option contains the
+                     link-layer address of the target.  It is used in
+                     Neighbor Advertisement and Redirect packets.
+
+                     These options MUST be silently ignored for other
+                     Neighbor Discovery messages.
+*/
+#[derive(Debug)]
+#[repr(C, packed)]
+pub struct LinkLayerAddressOption {
+    option_type: u8,
+    option_length: u8,
+    addr: MacAddress,
+}
+
+impl NdpOption for LinkLayerAddressOption {
+    fn get_type() -> NdpOptionType {
+        NdpOptionType::SourceLinkLayerAddress
+    }
+}
+
+/// Provides functions to implementors that allow the implementor to scan their NDP message for
+/// NDP options.  The type of options available are specific to the type of NDP Message
+/// Ensures that implementors support NdpMessageContents, which further enforces EndOffset + Sized
+/// which we require to do parsing
+pub trait NdpOptions: NdpMessageContents {
+    /// Attempts to lookup the source link layer address from the underlying NDP Message
+    fn get_source_link_layer_address_option(&self, payload_length: u16) -> Option<&MacAddress> {
+        let opt = self.option_scan::<LinkLayerAddressOption>(payload_length);
+        opt.map(|sll| &sll.addr)
+    }
+
+    /// Attempts to lookup the mtu option from the underlying NDP Message
+    fn get_mtu_option(&self, payload_length: u16) -> Option<u32> {
+        let mtu_option = self
+            .option_scan::<MtuOption>(payload_length)
+            .map(|mtu| mtu.get_mtu());
+        mtu_option
+    }
+
+    /// Searches through a packet for an option, returns the first it finds or none
+    /// The underlying type is determine by implementing the NdpOption trait which returns the NDP Option Type
+    fn option_scan<T: NdpOption>(&self, payload_length: u16) -> Option<&T> {
+        unsafe {
+            // the start of the options we offset from is the start of the ndp message
+            let opt_start = (self as *const Self) as *const u8;
+            let mut payload_offset = self.offset();
+            let requested_type = T::get_type();
+
+            // make sure we do not scan beyond the end of the packet
+            while (payload_offset as u16) < payload_length {
+                // we want to pull the first two bytes off, option type and option length
+                let seek_to = opt_start.add(payload_offset);
+                let option_meta = slice::from_raw_parts(seek_to, 2);
+
+                // first option meta is the option type, let's cast it to our NDPOptionType
+                let cur_type: NdpOptionType = FromPrimitive::from_u8(option_meta[0]).unwrap();
```

Comment:
Just occurred to me if this is not a known type, `unwrap` will panic here. It's safer to compare the other way, convert the enum to `u8`.  I think someone mentioned that they've seen off-spec options in live packets. Then this will blow up.

Reply:
fixed this. i added a check to see if the option is one of the Ndp options we defined. if it isnt, we set the ndp option typed to `Undefined`